src/arlo_e2e/io.py

```diff
@@ -0,0 +1,1148 @@
+# Twice as expensive per month for storage, but supports immediate deletion,
+# cheaper access. Suitable for testing, when we're creating and nuking these
+# files fairly quickly.
+
+# DEFAULT_S3_STORAGE_CLASS = "STANDARD_IA"
+# Suitable for 'Long-lived, infrequently accessed data'. Half the cost of "STANDARD"
+# for long-term storage, but you pay access charges, and objects have a minimum 30
+# day lifespan. Suitable for production deployment (?). Or maybe we want to use
+# "INTELLIGENT_TIERING" instead, which incurs a monthly cost, and will migrate
+# objects back and forth between these two classes.
+
+# Relevant documentation:
+# https://docs.aws.amazon.com/AmazonS3/latest/userguide/storage-class-intro.html
+# https://aws.amazon.com/s3/pricing/
+
+
+def _s3_client() -> S3Client:
+    """Gets S3 client resource. Initializes and caches as necessary."""
+    global _s3_client_handle
+
+    if _s3_client_handle is None:
+        _s3_client_handle = boto3.client("s3")
+    return _s3_client_handle
+
+
+def _md5_hash(b: bytes) -> str:
+    """Used for error-checking on S3 put_object calls."""
+    m = hashlib.md5()
+    m.update(b)
+    return b64encode(m.digest()).decode("utf-8")
+
+
+class FileRef(ABC):
+    """
+    Base class for LocalFileRef and S3FileRef. Never create one of these directly.
+    Instead use `make_file_ref`.
+
+    Rather than Python Path/PurePath objects, or whatever else, we're going to use this
+    FileRef class to capture what we're doing, keeping separate track of the root directory,
+    the list of subdirectories below that, and ultimately the file name.
+
+    Methods on this class deal with loading and saving files from text strings, arrays
+    of bytes and ElectionGuard `Serializable` objects. There's also special handling
+    for `CiphertextBallot` objects, where the FileRef instance should point to the
+    the root, and everything below that (e.g, `ballots/b0000/b00001234.json`) is
+    dealt with internally.
+
+    Subclasses of this will deal with local filesystems or with S3.
+    """
+
+    root_dir: str
+    subdirectories: List[str]
+    file_name: str
+
+    def __init__(self, root_dir: str, subdirectories: List[str], file_name: str):
+        """Don't use this constructor. Use `make_file_ref` instead."""
+        self.root_dir = root_dir
+        self.subdirectories = subdirectories
+        self.file_name = file_name
+
+    @abstractmethod
+    def update(
+        self,
+        new_file_name: str = None,
+        new_root_dir: str = None,
+        new_subdirs: List[str] = None,
+    ) -> "FileRef":
+        """
+        Makes a copy of the current FileRef with the opportunity to update the requested fields
+        to be a new file name, root directory, and/or subdirectory list. The original doesn't
+        mutate.
+        """
+        ...
+
+    def is_file(self) -> bool:
+        """Returns whether this is a file (True) or a directory (False)."""
+        return self.file_name != ""
```

Comment:
This is my internal representation. I leave all the directory names in the "subdirectories" field, and if the file name is empty, then I know I'm naming a directory rather than a file. That's the idea, anyway.